lib/spack/spack/fetch_strategy.py

```diff
@@ -158,64 +194,65 @@ class FetchStrategyComposite(object):
     Implements the GoF composite pattern.
     """
     matches = FetchStrategy.matches
-    set_stage = FetchStrategy.set_stage
 
 
+@fetcher
 class URLFetchStrategy(FetchStrategy):
     """FetchStrategy that pulls source code from a URL for an archive,
        checks the archive against a checksum,and decompresses the archive.
     """
-    enabled = True
     required_attributes = ['url']
+    #: Overridden by instances to set the curl executable
+    _curl = None
 
-    def __init__(self, url=None, digest=None, **kwargs):
-        super(URLFetchStrategy, self).__init__()
-
-        # If URL or digest are provided in the kwargs, then prefer
-        # those values.
-        self.url = kwargs.get('url', None)
-        if not self.url:
-            self.url = url
-
-        self.digest = next((kwargs[h] for h in crypto.hashes if h in kwargs),
-                           None)
+    def __init__(self, url, digest=None, **kwargs):
+        self.url = url
+        self.digest = next(
+            (kwargs[h] for h in crypto.hashes if h in kwargs), None
+        )
         if not self.digest:
             self.digest = digest
 
-        self.expand_archive = kwargs.get('expand', True)
         self.extra_curl_options = kwargs.get('curl_options', [])
-        self._curl = None
-
         self.extension = kwargs.get('extension', None)
 
-        if not self.url:
-            raise ValueError("URLFetchStrategy requires a url for fetching.")
+    @property
+    def archive_basename(self):
+        return os.path.basename(self.url)
+
+    @property
+    def expected_archive_files(self):
+        if self._expected_archive_files is None:
+            return [self.archive_basename]
+        return super(URLFetchStrategy, self).expected_archive_files
+
+    @expected_archive_files.setter
+    def expected_archive_files(self, value):
+        self._expected_archive_files = value
 
     @property
     def curl(self):
         if not self._curl:
             self._curl = which('curl', required=True)
         return self._curl
 
-    @_needs_stage
-    def fetch(self):
-        if self.archive_file:
-            tty.msg("Already downloaded %s" % self.archive_file)
-            return
+    def fetch(self, validate=True, expanded_source_tree=any):
```

Comment:
why does adding parameters like this simplify the logic?  I'm not really seeing it -- to me it makes the API harder to implement and less clearly divided into functions.  Does this help with the interaction between a fetcher and the stage somehow?